### ISSUE TITLE
sql: result type changes based on disk spilling

### ISSUE THREAD

**Issue by mjibson** [cockroachdb/cockroach#38985]
```
CREATE TABLE customers (name STRING PRIMARY KEY, id INT8, INDEX (id));

SELECT
    ix.indkey AS indkey
FROM
    pg_class AS t, pg_class AS i, pg_index AS ix, pg_attribute AS a
WHERE
    t.oid = ix.indrelid
    AND i.oid = ix.indexrelid
    AND a.attrelid = t.oid
    AND t.relkind = 'r'
    AND t.relname = 'customers'
GROUP BY
    i.relname, ix.indexrelid, ix.indisprimary, ix.indisunique, ix.indkey
ORDER BY
    i.relname;

SET CLUSTER SETTING "sql.distsql.temp_storage.workmem" = '1B';

SELECT
    ix.indkey AS indkey
FROM
    pg_class AS t, pg_class AS i, pg_index AS ix, pg_attribute AS a
WHERE
    t.oid = ix.indrelid
    AND i.oid = ix.indexrelid
    AND a.attrelid = t.oid
    AND t.relkind = 'r'
    AND t.relname = 'customers'
GROUP BY
    i.relname, ix.indexrelid, ix.indisprimary, ix.indisunique, ix.indkey
ORDER BY
    i.relname;
```

```
  indkey  
+--------+
       2  
       1  
(2 rows)

  indkey  
+--------+
  {2}     
  {1}     
(2 rows)
```

This can be reproduced in demo mode, but was found while trying to run the `fakedist-disk/orms`  logic test (a configuration that is not present, but I tried testing it). The same query is run twice, but with the workmem set to 1 byte. This causes the result type to change. Yeah.

**Comment by jordanlewis** [comment]
The problem here is that the very special compatibility type `int2vector`, which is identical to an integer array except for its formatting, doesn't roundtrip through the disk containers - it loses its oid wrapping and gets turned back into an ordinary int array.

I think this is a real bug but I'd say the priority is rather low.

**Comment by jordanlewis** [comment]
Oh, and I forgot to mention that `int2vector` is used in exactly 2 places: `pg_catalog.pg_index` and `pg_catalog.pg_trigger`. Users don't use this type in their tables.

**Comment by knz** [comment]
this has been fixed as of 21.1
